File: data-structures/binary-search.py
def binary_search(nums:list, target:int )->int:
  left = 0
  right = len(nums)-1

  return bs(nums,target, left, right)

def bs(nums:list, target:int,left:int, right:int)-> int:
  if(left>right):
    return -1
  else:
    piv = ((right+left) // 2)
    if(nums[piv] == target):
      return piv
    else:
        if(target<nums[piv]):
          return bs(nums, target,left,piv-1)
        else:
          return bs(nums,target, piv+1, right)


if __name__ == "__main__":
  res = binary_search([-1,0,3,5,9,12],2)

  print(res)
  # 0 1 2 3 4 5 | 6 7 8 | 9 10 |

File: data-structures/binary-search.cpp
#include <iostream>
#include <vector>

int binarySearch(std::vector<int>& nums, int target){
    int left = nums.begin();
    int right = nums.end() - 1;
    
    while(left<=right){
        auto mid = left + (right-left)/2;
        if(*mid == target){
            return mid;
        }else if (*mid<target){
            left = mid+1;
        }else{
            right = mid-1;
        }
    }
    
    
    return -1;
}

int main() {
    
    int arr[] = {0,4,5,7,11};
    
    int pos = binarySearch(arr,11);
    std::cout << pos << std::endl;
    return 0;
}
